File: src/main/java/pl/training/module8/collaboration/ChangeSet.java
package pl.training.module8.collaboration;

import java.util.List;
import java.util.Objects;
import java.util.Set;

public record ChangeSet(
        String title,
        Set<ChangeIntent> intents,
        List<VerificationEvidence> verificationEvidence,
        boolean independentlyGreenBuild) {

    public ChangeSet {
        Objects.requireNonNull(title, "title");
        if (title.isBlank()) {
            throw new IllegalArgumentException("title must not be blank");
        }
        intents = Set.copyOf(Objects.requireNonNull(intents, "intents"));
        verificationEvidence = List.copyOf(Objects.requireNonNull(
                verificationEvidence,
                "verificationEvidence"));
    }
}

File: csharp/src/Training.Module8/Collaboration/ChangeSet.cs
using System.Collections.Immutable;

namespace Training.Module8.Collaboration;

public sealed record ChangeSet
{
    public ChangeSet(
        string title,
        IEnumerable<ChangeIntent> intents,
        IEnumerable<VerificationEvidence> verificationEvidence,
        bool independentlyGreenBuild)
    {
        ArgumentNullException.ThrowIfNull(title);
        if (string.IsNullOrWhiteSpace(title))
        {
            throw new ArgumentException("title must not be blank");
        }
        ArgumentNullException.ThrowIfNull(intents);
        ArgumentNullException.ThrowIfNull(verificationEvidence);

        Title = title;
        Intents = intents.ToImmutableHashSet();
        VerificationEvidence = [.. verificationEvidence];
        IndependentlyGreenBuild = independentlyGreenBuild;
    }

    public string Title { get; }

    public IImmutableSet<ChangeIntent> Intents { get; }

    public ImmutableArray<VerificationEvidence> VerificationEvidence { get; }

    public bool IndependentlyGreenBuild { get; }
}
